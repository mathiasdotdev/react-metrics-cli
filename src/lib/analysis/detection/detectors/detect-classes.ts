/**
 * Détecteur de classes
 */

import { Declaration, DeclarationType } from '$types/analysis';
import { DetectorConfig } from '$types/config';
import { shouldIgnoreFile } from '@annotation/annotation-analyzer';
import { createDeclaration } from '@utils/declaration-utils';
import { isImportOrExport, isInString } from '@utils/string-utils';

/**
 * Détecte les déclarations de classes dans les lignes de code
 */
export const detectClasses = (
  filePath: string,
  lines: string[],
  config?: DetectorConfig,
): Declaration[] => {
  // Vérifier si le fichier entier doit être ignoré
  if (config?.ignoreAnnotations && shouldIgnoreFile(lines)) {
    return [];
  }

  const declarations: Declaration[] = [];
  const regex = /class\s+([a-zA-Z_$][a-zA-Z0-9_$]*)/g;

  for (let lineIndex = 0; lineIndex < lines.length; lineIndex++) {
    const line = lines[lineIndex];
    if (!line) continue;

    // Ignorer les commentaires
    const trimmed = line.trim();
    if (
      trimmed.startsWith('//') ||
      trimmed.startsWith('/*') ||
      trimmed.startsWith('*')
    ) {
      continue;
    }

    // Ignorer les imports/exports
    if (isImportOrExport(line)) {
      continue;
    }

    regex.lastIndex = 0;

    let match: RegExpExecArray | null;
    while ((match = regex.exec(line)) !== null) {
      const matchPosition = match.index;
      const name = match[1];

      if (name && !isInString(line, matchPosition)) {
        const column = line.indexOf(name) + 1;
        const lineNumber = lineIndex + 1;
        declarations.push(
          createDeclaration(
            name,
            DeclarationType.CLASS,
            filePath,
            lineNumber,
            column,
            undefined,
          ),
        );
      }
    }
  }

  return declarations;
};
