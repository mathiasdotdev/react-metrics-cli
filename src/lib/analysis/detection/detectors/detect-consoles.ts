/**
 * Détecteur de console.log et autres méthodes console
 */

import { shouldIgnoreFile } from '@annotation/annotation-analyzer';
import { Declaration, DeclarationType } from '$types/analysis';
import { DetectorConfig } from '$types/config';
import { createDeclaration } from '@utils/declaration-utils';
import { isInString } from '@utils/string-utils';

/**
 * Détecte les appels console
 */
export const detectConsoles = (
  filePath: string,
  lines: string[],
  config?: DetectorConfig,
): Declaration[] => {
  // Vérifier si le fichier entier doit être ignoré
  if (config?.ignoreAnnotations && shouldIgnoreFile(lines)) {
    return [];
  }

  const declarations: Declaration[] = [];
  const consoleRegex =
    /\b(console\.(log|warn|error|info|debug|table|trace|dir|dirxml|group|groupEnd|time|timeEnd|assert|count|countReset))\s*\(/g;

  for (let lineIndex = 0; lineIndex < lines.length; lineIndex++) {
    const line = lines[lineIndex];
    if (!line) continue;

    // Ignorer les commentaires
    const trimmed = line.trim();
    if (
      trimmed.startsWith('//') ||
      trimmed.startsWith('/*') ||
      trimmed.startsWith('*')
    ) {
      continue;
    }

    const lineNumber = lineIndex + 1;

    consoleRegex.lastIndex = 0;
    let match: RegExpExecArray | null;
    while ((match = consoleRegex.exec(line)) !== null) {
      const matchPosition = match.index;
      const fullMatch = match[1];

      if (fullMatch && !isInString(line, matchPosition)) {
        const column = matchPosition + 1;
        const declaration = createDeclaration(
          fullMatch,
          DeclarationType.CONSOLE,
          filePath,
          lineNumber,
          column,
        );
        declaration.context = fullMatch;
        declarations.push(declaration);
      }
    }
  }

  return declarations;
};
