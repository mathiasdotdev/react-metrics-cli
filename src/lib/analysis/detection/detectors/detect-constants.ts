/**
 * Détecteur de constantes
 */

import { Declaration, DeclarationType } from '$types/analysis';
import { DetectorConfig } from '$types/config';
import {
  shouldIgnoreDeclarationWithContext,
  shouldIgnoreFile,
} from '@annotation/annotation-analyzer';
import { createDeclaration } from '@utils/declaration-utils';
import { isImportOrExport, isInString } from '@utils/string-utils';

/**
 * Détecte les déclarations de constantes dans les lignes de code
 */
export const detectConstants = (
  filePath: string,
  lines: string[],
  config?: DetectorConfig,
): Declaration[] => {
  // Vérifier si le fichier entier doit être ignoré
  if (config?.ignoreAnnotations && shouldIgnoreFile(lines)) {
    return [];
  }

  const declarations: Declaration[] = [];
  const regex = /\b(const|let|var)\s+([a-zA-Z_$][a-zA-Z0-9_$]*)/g;
  const destructuringRegex = /\b(const|let|var)\s*\{\s*([^}]+)\s*\}\s*=/g;
  const arrayDestructuringRegex = /\b(const|let|var)\s*\[\s*([^\]]+)\s*\]\s*=/g;

  for (let lineIndex = 0; lineIndex < lines.length; lineIndex++) {
    const line = lines[lineIndex];
    if (!line) continue;

    // Ignorer les commentaires
    const trimmed = line.trim();
    if (
      trimmed.startsWith('//') ||
      trimmed.startsWith('/*') ||
      trimmed.startsWith('*')
    ) {
      continue;
    }

    // Ignorer les imports/exports
    if (isImportOrExport(line)) {
      continue;
    }

    const lineNumber = lineIndex + 1;

    // Vérifier si la déclaration doit être ignorée
    if (
      config?.ignoreAnnotations &&
      shouldIgnoreDeclarationWithContext(lines, lineNumber)
    ) {
      continue;
    }

    // Détecter la destructuration d'objet: const { a, b } = ...
    destructuringRegex.lastIndex = 0;
    let match: RegExpExecArray | null;
    while ((match = destructuringRegex.exec(line)) !== null) {
      const matchPosition = match.index;
      if (!isInString(line, matchPosition) && match[2]) {
        const props = match[2].split(',');
        // On ne détecte que le premier pour éviter la duplication
        const firstProp = props[0]?.trim().split(/\s|:/)[0];
        if (firstProp) {
          const column = line.indexOf(firstProp) + 1;
          declarations.push(
            createDeclaration(
              firstProp,
              DeclarationType.CONSTANT,
              filePath,
              lineNumber,
              column,
              undefined,
            ),
          );
        }
      }
    }

    // Si on a déjà détecté une destructuration, passer à la ligne suivante
    if (declarations.some((d) => d.location.line === lineNumber)) {
      continue;
    }

    // Détecter la destructuration de tableau: const [a, b] = ...
    arrayDestructuringRegex.lastIndex = 0;
    while ((match = arrayDestructuringRegex.exec(line)) !== null) {
      const matchPosition = match.index;
      if (!isInString(line, matchPosition) && match[2]) {
        const items = match[2].split(',');
        // On ne détecte que le premier
        const firstItem = items[0]?.trim();
        if (firstItem) {
          const column = line.indexOf(firstItem) + 1;
          declarations.push(
            createDeclaration(
              firstItem,
              DeclarationType.CONSTANT,
              filePath,
              lineNumber,
              column,
              undefined,
            ),
          );
        }
      }
    }

    // Si on a déjà détecté une destructuration, passer à la ligne suivante
    if (declarations.some((d) => d.location.line === lineNumber)) {
      continue;
    }

    // Détecter les déclarations normales: const x = ...
    // Mais exclure celles avec annotation de type (const x: Type = ...) car elles sont détectées comme PROP
    regex.lastIndex = 0;
    while ((match = regex.exec(line)) !== null) {
      const matchPosition = match.index;
      const name = match[2];

      // Vérifier que le match n'est pas dans une chaîne
      if (name && !isInString(line, matchPosition)) {
        // Vérifier si c'est une constante avec annotation de type
        const afterName = line.substring(matchPosition + match[0].length);
        if (afterName.match(/^\s*:/)) {
          // C'est une constante typée, elle sera détectée comme PROP
          continue;
        }

        // Vérifier si c'est une arrow function
        if (
          afterName.match(/^\s*=\s*(\([^)]*\)|[a-zA-Z_$][a-zA-Z0-9_$]*)\s*=>/)
        ) {
          // C'est une arrow function, elle sera détectée comme FUNCTION
          continue;
        }

        const column = line.indexOf(name, matchPosition) + 1;
        declarations.push(
          createDeclaration(
            name,
            DeclarationType.CONSTANT,
            filePath,
            lineNumber,
            column,
            undefined,
          ),
        );
      }
    }
  }

  return declarations;
};
