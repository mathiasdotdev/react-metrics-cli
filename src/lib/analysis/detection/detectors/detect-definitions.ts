/**
 * Détecteur de types et interfaces TypeScript
 */

import { shouldIgnoreFile } from '@annotation/annotation-analyzer';
import { Declaration, DeclarationType } from '$types/analysis';
import { DetectorConfig } from '$types/config';
import { createDeclaration } from '@utils/declaration-utils';
import { isValidName } from '@utils/identifier-utils';
import { isInString } from '@utils/string-utils';

/**
 * Détecte les déclarations de types et interfaces dans les lignes de code
 */
export const detectDefinitions = (
  filePath: string,
  lines: string[],
  config?: DetectorConfig,
): Declaration[] => {
  // Vérifier si le fichier entier doit être ignoré
  if (config?.ignoreAnnotations && shouldIgnoreFile(lines)) {
    return [];
  }

  const declarations: Declaration[] = [];
  const regexType =
    /^export\s+type\s+([a-zA-Z_$][a-zA-Z0-9_$]*)|^type\s+([a-zA-Z_$][a-zA-Z0-9_$]*)/;
  const regexInterface =
    /^export\s+interface\s+([a-zA-Z_$][a-zA-Z0-9_$]*)|^interface\s+([a-zA-Z_$][a-zA-Z0-9_$]*)/;

  for (let lineIndex = 0; lineIndex < lines.length; lineIndex++) {
    const line = lines[lineIndex];
    if (!line) continue;

    const ligneTrimmed = line.trim();

    if (!ligneTrimmed) {
      continue;
    }

    const lineNumber = lineIndex + 1;

    // Détecter types
    const matchType = regexType.exec(ligneTrimmed);
    if (matchType) {
      const nom = matchType[1] || matchType[2];
      if (nom && isValidName(nom)) {
        const position = line.indexOf(nom);
        if (position !== -1 && !isInString(line, position)) {
          declarations.push(
            createDeclaration(
              nom,
              DeclarationType.DEFINITION,
              filePath,
              lineNumber,
              position + 1,
              'type',
            ),
          );
        }
      }
    }

    // Détecter interfaces
    const matchInterface = regexInterface.exec(ligneTrimmed);
    if (matchInterface) {
      const nom = matchInterface[1] || matchInterface[2];
      if (nom && isValidName(nom)) {
        const position = line.indexOf(nom);
        if (position !== -1 && !isInString(line, position)) {
          declarations.push(
            createDeclaration(
              nom,
              DeclarationType.DEFINITION,
              filePath,
              lineNumber,
              position + 1,
              'interface',
            ),
          );
        }
      }
    }
  }

  return declarations;
};
