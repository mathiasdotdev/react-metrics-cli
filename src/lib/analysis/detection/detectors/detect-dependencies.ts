/**
 * Détecteur de dépendances package.json
 */

import { Declaration, DeclarationType } from '$types/analysis';
import * as fs from 'fs/promises';
import * as path from 'path';

/**
 * Lit le contenu d'un fichier
 */
const readFileContent = async (filePath: string): Promise<string> => {
  try {
    const content = await fs.readFile(filePath, 'utf-8');
    return content;
  } catch (error) {
    throw new Error(`Impossible de lire le fichier ${filePath}: ${error}`);
  }
};

/**
 * Détecte les dépendances dans le package.json
 */
export const detectDependencies = async (
  projectPath: string,
): Promise<Declaration[]> => {
  const packageJsonPath = path.join(projectPath, 'package.json');

  try {
    const content = await readFileContent(packageJsonPath);
    const packageJson = JSON.parse(content);
    const declarations: Declaration[] = [];

    // Parcourir dependencies
    if (packageJson.dependencies) {
      const depsStart = content.indexOf('"dependencies"');
      if (depsStart !== -1) {
        for (const [depName, _version] of Object.entries(
          packageJson.dependencies,
        )) {
          // Trouver la ligne exacte de cette dépendance
          const depPattern = `"${depName}"`;
          const depIndex = content.indexOf(depPattern, depsStart);
          if (depIndex !== -1) {
            const beforeDep = content.substring(0, depIndex);
            const depLine = (beforeDep.match(/\n/g) || []).length + 1;

            declarations.push({
              name: depName,
              type: DeclarationType.DEPENDENCY,
              location: {
                filePath: packageJsonPath,
                line: depLine,
                column:
                  content.substring(beforeDep.lastIndexOf('\n') + 1, depIndex)
                    .length + 1,
              },
              isUsedLocally: false,
              isImportedExternally: false,
              context: 'dependency',
            });
          }
        }
      }
    }

    // Parcourir devDependencies
    if (packageJson.devDependencies) {
      const devDepsStart = content.indexOf('"devDependencies"');
      if (devDepsStart !== -1) {
        for (const [depName, _version] of Object.entries(
          packageJson.devDependencies,
        )) {
          const depPattern = `"${depName}"`;
          const depIndex = content.indexOf(depPattern, devDepsStart);
          if (depIndex !== -1) {
            const beforeDep = content.substring(0, depIndex);
            const depLine = (beforeDep.match(/\n/g) || []).length + 1;

            declarations.push({
              name: depName,
              type: DeclarationType.DEPENDENCY,
              location: {
                filePath: packageJsonPath,
                line: depLine,
                column:
                  content.substring(beforeDep.lastIndexOf('\n') + 1, depIndex)
                    .length + 1,
              },
              isUsedLocally: false,
              isImportedExternally: false,
              context: 'devDependency',
            });
          }
        }
      }
    }

    return declarations;
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
  } catch (_error) {
    return [];
  }
};
