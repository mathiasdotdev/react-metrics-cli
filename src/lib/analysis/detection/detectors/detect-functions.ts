/**
 * Détecteur de fonctions
 */

import { Declaration, DeclarationType } from '$types/analysis';
import { DetectorConfig } from '$types/config';
import {
  shouldIgnoreDeclarationWithContext,
  shouldIgnoreFile,
} from '@annotation/annotation-analyzer';
import { createDeclaration } from '@utils/declaration-utils';
import { isImportOrExport, isInString } from '@utils/string-utils';

/**
 * Détecte les déclarations de fonctions dans les lignes de code
 */
export const detectFunctions = (
  filePath: string,
  lines: string[],
  config?: DetectorConfig,
): Declaration[] => {
  const declarations: Declaration[] = [];

  // Vérifier si le fichier entier doit être ignoré
  if (config?.ignoreAnnotations && shouldIgnoreFile(lines)) {
    return declarations;
  }

  const functionRegex = /function\s+([a-zA-Z_$][a-zA-Z0-9_$]*)/g;
  const arrowRegex =
    /const\s+([a-zA-Z_$][a-zA-Z0-9_$]*)\s*=\s*(?:async\s*)?\([^)]*\)\s*(?::\s*[^=]+)?\s*=>/g;

  for (let lineIndex = 0; lineIndex < lines.length; lineIndex++) {
    const line = lines[lineIndex];
    if (!line) continue;

    // Ignorer les commentaires
    const trimmed = line.trim();
    if (
      trimmed.startsWith('//') ||
      trimmed.startsWith('/*') ||
      trimmed.startsWith('*')
    ) {
      continue;
    }

    // Ignorer les imports/exports
    if (isImportOrExport(line)) {
      continue;
    }

    const lineNumber = lineIndex + 1;

    // Détecter les fonctions classiques
    functionRegex.lastIndex = 0;
    let match: RegExpExecArray | null;
    while ((match = functionRegex.exec(line)) !== null) {
      const matchPosition = match.index;
      const name = match[1];

      if (name && !isInString(line, matchPosition)) {
        // Vérifier les annotations d'ignore
        if (
          config?.ignoreAnnotations &&
          shouldIgnoreDeclarationWithContext(
            lines,
            lineNumber,
            config.contextLines || 3,
          )
        ) {
          continue;
        }

        const column = line.indexOf(name) + 1;
        declarations.push(
          createDeclaration(
            name,
            DeclarationType.FUNCTION,
            filePath,
            lineNumber,
            column,
          ),
        );
      }
    }

    // Détecter les fonctions fléchées
    arrowRegex.lastIndex = 0;
    while ((match = arrowRegex.exec(line)) !== null) {
      const matchPosition = match.index;
      const name = match[1];

      if (name && !isInString(line, matchPosition)) {
        // Vérifier les annotations d'ignore
        if (
          config?.ignoreAnnotations &&
          shouldIgnoreDeclarationWithContext(
            lines,
            lineNumber,
            config.contextLines || 3,
          )
        ) {
          continue;
        }

        const column = line.indexOf(name) + 1;
        declarations.push(
          createDeclaration(
            name,
            DeclarationType.FUNCTION,
            filePath,
            lineNumber,
            column,
          ),
        );
      }
    }

    // NE PAS détecter les méthodes de classe (constructor, render, etc.)
    // Les méthodes font partie de la classe et ne sont jamais "mortes" indépendamment
  }

  return declarations;
};
