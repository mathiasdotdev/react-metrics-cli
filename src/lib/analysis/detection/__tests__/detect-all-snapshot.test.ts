import { describe, expect, test } from 'bun:test';
import {
  createTestCase,
  runDetectionSnapshot,
} from './helpers/detectionSnapshotHelper';
import {
  CLASS_SAMPLES,
  CONSOLE_SAMPLES,
  CONSTANT_SAMPLES,
  DEFINITION_SAMPLES,
  DEPENDENCY_SAMPLES,
  DEPRECATED_SAMPLES,
  EXPORT_SAMPLES,
  FUNCTION_SAMPLES,
  PROP_SAMPLES,
} from '$tests/fixtures/code-samples';

// Import detectors
import { detectClasses } from '../detectors/detect-classes';
import { detectConsoles } from '../detectors/detect-consoles';
import { detectConstants } from '../detectors/detect-constants';
import { detectDefinitions } from '../detectors/detect-definitions';
import { detectDependencies } from '../detectors/detect-dependencies';
import { detectDeprecated } from '../detectors/detect-deprecated';
import { detectExports } from '../detectors/detect-exports';
import { detectFunctions } from '../detectors/detect-functions';
import { detectProps } from '../detectors/detect-props';

describe('Detection - All Detectors Snapshot Tests', () => {
  const defaultConfig = { ignoreAnnotations: false };

  test('detectFunctions: all cases', () => {
    const testCases: Record<string, { code: string; config?: any }> = {};

    Object.entries(FUNCTION_SAMPLES).forEach(([key, code]) => {
      testCases[key] = createTestCase(code, defaultConfig);
    });

    const result = runDetectionSnapshot(detectFunctions, testCases);
    expect(result).toMatchSnapshot();
  });

  test('detectClasses: all cases', () => {
    const testCases: Record<string, { code: string; config?: any }> = {};

    Object.entries(CLASS_SAMPLES).forEach(([key, code]) => {
      testCases[key] = createTestCase(code);
    });

    const result = runDetectionSnapshot(detectClasses, testCases);
    expect(result).toMatchSnapshot();
  });

  test('detectConstants: all cases', () => {
    const testCases: Record<string, { code: string; config?: any }> = {};

    Object.entries(CONSTANT_SAMPLES).forEach(([key, code]) => {
      testCases[key] = createTestCase(code);
    });

    const result = runDetectionSnapshot(detectConstants, testCases);
    expect(result).toMatchSnapshot();
  });

  test('detectProps: all cases', () => {
    const testCases: Record<string, { code: string; config?: any }> = {};

    Object.entries(PROP_SAMPLES).forEach(([key, code]) => {
      testCases[key] = createTestCase(code, defaultConfig);
    });

    const result = runDetectionSnapshot(detectProps, testCases);
    expect(result).toMatchSnapshot();
  });

  test('detectDefinitions: all cases', () => {
    const testCases: Record<string, { code: string; config?: any }> = {};

    Object.entries(DEFINITION_SAMPLES).forEach(([key, code]) => {
      testCases[key] = createTestCase(code, defaultConfig);
    });

    const result = runDetectionSnapshot(detectDefinitions, testCases);
    expect(result).toMatchSnapshot();
  });

  test('detectExports: all cases', () => {
    const testCases: Record<string, { code: string; config?: any }> = {};

    Object.entries(EXPORT_SAMPLES).forEach(([key, code]) => {
      testCases[key] = createTestCase(code, defaultConfig);
    });

    const result = runDetectionSnapshot(detectExports, testCases);
    expect(result).toMatchSnapshot();
  });

  test('detectConsoles: all cases', () => {
    const testCases: Record<string, { code: string; config?: any }> = {};

    Object.entries(CONSOLE_SAMPLES).forEach(([key, code]) => {
      testCases[key] = createTestCase(code);
    });

    const result = runDetectionSnapshot(detectConsoles, testCases);
    expect(result).toMatchSnapshot();
  });

  test('detectDependencies: all cases', async () => {
    const results: string[] = [];

    // Trier les cas de test par clé
    const sortedCases = Object.keys(DEPENDENCY_SAMPLES).sort();

    for (const caseName of sortedCases) {
      const packageJsonContent =
        DEPENDENCY_SAMPLES[caseName as keyof typeof DEPENDENCY_SAMPLES];

      // detectDependencies accepte maintenant le contenu JSON directement pour les tests
      let declarations: any[] = [];
      try {
        const packageJson = JSON.parse(packageJsonContent);
        const allDeps = {
          ...packageJson.dependencies,
          ...packageJson.devDependencies,
        };

        declarations = Object.keys(allDeps || {}).map((depName, index) => ({
          name: depName,
          type: 7, // DeclarationType.DEPENDENCY
          location: {
            filePath: '/test/package.json',
            line: index + 3,
            column: 5,
          },
          isUsedLocally: false,
          isImportedExternally: false,
        }));
      } catch {
        // Invalid JSON
        declarations = [];
      }

      results.push(`\n=== ${caseName} ===`);

      if (declarations.length === 0) {
        results.push('(aucune dépendance détectée)');
      } else {
        // Trier par nom de package
        const sorted = declarations.sort((a, b) =>
          a.name.localeCompare(b.name),
        );
        results.push(`Count: ${declarations.length}`);
        sorted.forEach((decl) => {
          results.push(
            `DEPENDENCY | ${decl.name} | package.json:${decl.location.line}:${decl.location.column}`,
          );
        });
      }
    }

    expect(results.join('\n')).toMatchSnapshot();
  });

  test('detectDeprecated: all cases', () => {
    const testCases: Record<string, { code: string; config?: any }> = {};

    Object.entries(DEPRECATED_SAMPLES).forEach(([key, code]) => {
      testCases[key] = createTestCase(code, defaultConfig);
    });

    const result = runDetectionSnapshot(detectDeprecated, testCases);
    expect(result).toMatchSnapshot();
  });
});
