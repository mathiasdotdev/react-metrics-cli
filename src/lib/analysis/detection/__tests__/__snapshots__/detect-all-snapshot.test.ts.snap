// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`Detection - All Detectors Snapshot Tests detectFunctions: all cases 1`] = `
"
=== arrowFunction ===
Count: 1
FUNCTION | unusedFunction | file.ts:1:7

=== arrowFunctionWithParams ===
Count: 1
FUNCTION | multiply | file.ts:1:7

=== asyncFunction ===
Count: 1
FUNCTION | fetchData | file.ts:1:16

=== classicFunction ===
Count: 1
FUNCTION | divide | file.ts:1:10

=== deprecatedArrowFunction ===
Count: 1
FUNCTION | legacyHandler | file.ts:2:7

=== deprecatedFunction ===
Count: 1
FUNCTION | oldFunction | file.ts:4:10

=== exportedArrowFunction ===
(aucune déclaration détectée)

=== exportedFunction ===
(aucune déclaration détectée)

=== functionInComment ===
(aucune déclaration détectée)

=== functionInString ===
(aucune déclaration détectée)

=== functionWithDollarSign ===
Count: 1
FUNCTION | $jqueryFunction | file.ts:1:7

=== functionWithUnderscore ===
Count: 1
FUNCTION | _privateFunction | file.ts:1:10

=== multipleFunctions ===
Count: 3
FUNCTION | first | file.ts:1:10
FUNCTION | second | file.ts:3:7
FUNCTION | third | file.ts:5:10"
`;

exports[`Detection - All Detectors Snapshot Tests detectClasses: all cases 1`] = `
"
=== abstractClass ===
Count: 1
CLASS | AbstractClass | file.ts:1:16

=== basicClass ===
Count: 1
CLASS | MyClass | file.ts:1:7

=== classInComment ===
(aucune déclaration détectée)

=== classInString ===
(aucune déclaration détectée)

=== classWithExtends ===
Count: 1
CLASS | ChildClass | file.ts:1:7

=== classWithImplements ===
Count: 1
CLASS | MyClass | file.ts:1:7

=== classWithMultipleImplements ===
Count: 1
CLASS | MyClass | file.ts:1:7

=== defaultExportedClass ===
(aucune déclaration détectée)

=== deprecatedClass ===
Count: 1
CLASS | OldClass | file.ts:4:7

=== exportedClass ===
(aucune déclaration détectée)

=== genericClass ===
Count: 1
CLASS | GenericClass | file.ts:1:7

=== reactComponentClass ===
Count: 1
CLASS | MyComponent | file.ts:1:7"
`;

exports[`Detection - All Detectors Snapshot Tests detectConstants: all cases 1`] = `
"
=== arrayDestructured ===
Count: 1
CONSTANT | first | file.ts:1:8

=== constDeclaration ===
Count: 1
CONSTANT | myConstant | file.ts:1:7

=== constInComment ===
(aucune déclaration détectée)

=== constWithDollarSign ===
Count: 1
CONSTANT | $jqueryStyle | file.ts:1:7

=== constWithUnderscore ===
Count: 1
CONSTANT | _privateConstant | file.ts:1:7

=== deprecatedConstant ===
Count: 1
CONSTANT | OLD_CONSTANT | file.ts:4:7

=== destructured ===
Count: 1
CONSTANT | name | file.ts:1:9

=== exportedConstant ===
(aucune déclaration détectée)

=== letDeclaration ===
Count: 1
CONSTANT | myVariable | file.ts:1:5

=== multipleConstants ===
Count: 4
CONSTANT | first | file.ts:1:7
CONSTANT | second | file.ts:2:7
CONSTANT | third | file.ts:3:5
CONSTANT | fourth | file.ts:4:5

=== uppercaseConstant ===
Count: 1
CONSTANT | MAX_SIZE | file.ts:1:7

=== varDeclaration ===
Count: 1
CONSTANT | oldStyleVar | file.ts:1:5"
`;

exports[`Detection - All Detectors Snapshot Tests detectProps: all cases 1`] = `
"
=== deprecatedProp ===
Count: 1
PROP | oldValue | file.ts:2:7

=== inlineProperty ===
Count: 2
PROP | name | file.ts:1:16
PROP | age | file.ts:1:30

=== multipleProp ===
Count: 3
PROP | host | file.ts:2:3
PROP | port | file.ts:3:3
PROP | ssl | file.ts:4:3

=== objectLiteral ===
Count: 2
PROP | key | file.ts:2:3
PROP | timeout | file.ts:3:3

=== propInInterface ===
Count: 2
PROP | name | file.ts:2:3
PROP | email | file.ts:3:3

=== propInType ===
Count: 2
PROP | name | file.ts:2:3
PROP | age | file.ts:3:3

=== typedConstant ===
Count: 1
PROP | API_URL | file.ts:1:7"
`;

exports[`Detection - All Detectors Snapshot Tests detectDefinitions: all cases 1`] = `
"
=== conditionalType ===
Count: 1
DEFINITION | IsString | file.ts:1:6

=== deprecatedType ===
Count: 1
DEFINITION | OldType | file.ts:4:6

=== exportedInterface ===
Count: 1
DEFINITION | IConfig | file.ts:1:18

=== exportedType ===
Count: 1
DEFINITION | Status | file.ts:1:13

=== genericInterface ===
Count: 1
DEFINITION | IResponse | file.ts:1:11

=== genericType ===
Count: 1
DEFINITION | Result | file.ts:1:6

=== interfaceDeclaration ===
Count: 1
DEFINITION | IUser | file.ts:1:11

=== interfaceWithExtends ===
Count: 1
DEFINITION | IChild | file.ts:1:11

=== mappedType ===
Count: 1
DEFINITION | Readonly | file.ts:1:6

=== typeDeclaration ===
Count: 1
DEFINITION | MyType | file.ts:1:6

=== typeInComment ===
(aucune déclaration détectée)

=== typeWithIntersection ===
Count: 1
DEFINITION | Combined | file.ts:1:6

=== typeWithUnion ===
Count: 1
DEFINITION | Status | file.ts:1:6"
`;

exports[`Detection - All Detectors Snapshot Tests detectExports: all cases 1`] = `
"
=== deprecatedExport ===
Count: 1
EXPORT | OLD_API | file.ts:2:14

=== exportArrowFunction ===
Count: 1
EXPORT | handler | file.ts:1:14

=== exportAsyncFunction ===
Count: 1
EXPORT | fetchData | file.ts:1:23

=== exportClass ===
Count: 1
EXPORT | Service | file.ts:1:14

=== exportConst ===
Count: 1
EXPORT | API_KEY | file.ts:1:14

=== exportDefault ===
Count: 1
EXPORT | MyComponent | file.ts:1:16

=== exportDefaultClass ===
Count: 1
EXPORT | class | file.ts:1:16

=== exportDefaultFunction ===
Count: 1
EXPORT | function | file.ts:1:16

=== exportDestructuring ===
Count: 2
EXPORT | a | file.ts:1:16
EXPORT | b | file.ts:1:19

=== exportFunction ===
Count: 1
EXPORT | calculate | file.ts:1:17

=== exportInComment ===
(aucune déclaration détectée)

=== exportReactComponent ===
Count: 1
EXPORT | Button | file.ts:1:14

=== exportWithAlias ===
Count: 1
EXPORT | myFunction | file.ts:1:10

=== exportWithTypeAnnotation ===
Count: 1
EXPORT | value | file.ts:1:14

=== multipleNamedExports ===
Count: 3
EXPORT | fn1 | file.ts:1:10
EXPORT | fn2 | file.ts:1:15
EXPORT | fn3 | file.ts:1:20

=== namedExport ===
Count: 1
EXPORT | myFunction | file.ts:1:10"
`;

exports[`Detection - All Detectors Snapshot Tests detectConsoles: all cases 1`] = `
"
=== consoleDebug ===
Count: 1
CONSOLE | console.debug | file.ts:1:1

=== consoleError ===
Count: 1
CONSOLE | console.error | file.ts:1:1

=== consoleInArrowFunction ===
Count: 1
CONSOLE | console.log | file.ts:1:19

=== consoleInComment ===
(aucune déclaration détectée)

=== consoleInConditional ===
Count: 1
CONSOLE | console.log | file.ts:2:3

=== consoleInString ===
(aucune déclaration détectée)

=== consoleInTryCatch ===
Count: 1
CONSOLE | console.error | file.ts:4:3

=== consoleInfo ===
Count: 1
CONSOLE | console.info | file.ts:1:1

=== consoleLog ===
Count: 1
CONSOLE | console.log | file.ts:1:1

=== consoleWarn ===
Count: 1
CONSOLE | console.warn | file.ts:1:1

=== consoleWithComplexArgs ===
Count: 1
CONSOLE | console.log | file.ts:1:1

=== consoleWithMethodChaining ===
Count: 1
CONSOLE | console.log | file.ts:1:1

=== multipleConsoleCalls ===
Count: 3
CONSOLE | console.log | file.ts:1:1
CONSOLE | console.error | file.ts:2:1
CONSOLE | console.warn | file.ts:3:1"
`;

exports[`Detection - All Detectors Snapshot Tests detectDependencies: all cases 1`] = `
"
=== bothDependencies ===
Count: 2
DEPENDENCY | express | package.json:3:5
DEPENDENCY | nodemon | package.json:4:5

=== devDependencies ===
Count: 2
DEPENDENCY | typescript | package.json:3:5
DEPENDENCY | vitest | package.json:4:5

=== invalidJson ===
(aucune dépendance détectée)

=== manyDependencies ===
Count: 5
DEPENDENCY | dep1 | package.json:3:5
DEPENDENCY | dep2 | package.json:4:5
DEPENDENCY | dep3 | package.json:5:5
DEPENDENCY | dep4 | package.json:6:5
DEPENDENCY | dep5 | package.json:7:5

=== noDependencies ===
(aucune dépendance détectée)

=== scopedPackages ===
Count: 2
DEPENDENCY | @angular/core | package.json:4:5
DEPENDENCY | @types/node | package.json:3:5

=== simpleDependencies ===
Count: 2
DEPENDENCY | axios | package.json:4:5
DEPENDENCY | react | package.json:3:5"
`;

exports[`Detection - All Detectors Snapshot Tests detectDeprecated: all cases 1`] = `
"
=== deprecatedArrowFunction ===
Count: 1
FUNCTION | legacyHandler | file.ts:2:7

=== deprecatedClass ===
Count: 1
CLASS | OldClass | file.ts:4:7

=== deprecatedConstant ===
Count: 1
CONSTANT | OLD_VALUE | file.ts:2:7

=== deprecatedExport ===
Count: 1
CONSTANT | OLD_API | file.ts:4:14

=== deprecatedFunction ===
Count: 1
FUNCTION | oldFunction | file.ts:4:10

=== deprecatedInterface ===
Count: 1
DEFINITION | IOldInterface | file.ts:4:11

=== deprecatedProp ===
Count: 1
PROP | oldField | file.ts:3:3

=== deprecatedType ===
Count: 1
DEFINITION | OldType | file.ts:2:6

=== multipleDeprecated ===
Count: 2
CONSTANT | old1 | file.ts:2:7
FUNCTION | old2 | file.ts:7:10

=== nonDeprecated ===
(aucune déclaration détectée)"
`;
